swarm/pss/pss.go

```diff
@@ -644,7 +663,48 @@ func (self *Pss) send(to []byte, topic Topic, msg []byte, asymmetric bool, key [
 		Expire:  uint32(time.Now().Add(self.msgTTL).Unix()),
 		Payload: envelope,
 	}
-	return self.forward(pssmsg)
+	return self.enqueue(pssmsg)
+}
+
+func (self *Pss) enqueue(msg *PssMsg) error {
+	// check if message is expired
+	if self.isMsgExpired(msg) {
+		log.Trace("pss expired :/ ... dropping")
+		return errors.New("expired")
+	}
+	// flood guard:
+	// don't allow identical messages we saw shortly before
+	// cache the message
+	digest, err := self.storeMsg(msg)
+	if err != nil {
+		log.Warn(fmt.Sprintf("could not store message %v to cache: %v", msg, err))
+	}
+	if self.checkFwdCache(nil, digest) {
+		log.Trace(fmt.Sprintf("pss relay block-cache match: FROM %x TO %x", self.Overlay.BaseAddr(), common.ToHex(msg.To)))
+		return nil
+	}
+
+	self.outboxMu.Lock()
+	defer self.outboxMu.Unlock()
+	nextPos := (self.outboxLastCursor + 1) % defaultOutboxQueueSize
+	if nextPos == self.outboxFirstCursor {
+		return errors.New("outbox full")
+	}
+	self.outbox[self.outboxLastCursor] = msg
+	self.outboxLastCursor++
+	return nil
+}
+
+func (self *Pss) dequeue() {
+	if self.outboxFirstCursor == self.outboxLastCursor {
+		time.Sleep(time.Millisecond * 100)
+		return
+	}
+	msg := self.outbox[self.outboxFirstCursor]
```

Comment:
I am pretty sure that slices are not safe even for concurrent read. Shouldn't this be locked with the `outboxMu`?

Reply:
But it's not a slice, it's a *PssMsg.

Reply:
```
outbox            []*PssMsg
```

You are reading from the `outbox` slice in `dequeue()`, while at the same time `outbox` could be getting modified in `enqueue` (in another goroutine).

Reply:
Ah. They're called slices not arrays maybe in Go? Of course, this is not a map.

So, I'm not sure if this is a problem here. The array size will not be changed, and the data is just a pointer, whose instruction will be a single write anyway, I assume?

Also, it will only write if the `First` and `Last` cursor is not the same value (thus pointing to the same array element). The `First` cursor is only incremented after the read.

What do you think?

Reply:
Yes, in Go you have `slices`, `arrays` and `maps`. `slice` is like an `array` but a bit different.

Both `slices` and `maps` are not safe for concurrent use, even when you are only reading from them. You can see it reported here:

```
WARNING: DATA RACE
Read at 0x00c420892000 by goroutine 38:
  github.com/ethereum/go-ethereum/swarm/pss.(*Pss).dequeue()
      /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/pss/pss.go:703 +0xc7
  github.com/ethereum/go-ethereum/swarm/pss.(*Pss).Start.func2()
      /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/pss/pss.go:186 +0x15a

Previous write at 0x00c420892000 by goroutine 197:
  [failed to restore the stack]
```

Anytime you have more than one goroutine accessing a slice or a map, you have to lock the slice or the map. I think the only exception is if you KNOW that you are going to be reading different indices within a slice (for map this doesn't apply).

Bottom line - better lock any slice or map that you use concurrently :)

Reply:
What is the traditional equivalent of go slice? Is it analogous to a dynamic array? As in, are all dynamic arrays in go slices, and all static arrays arrays?

Reply:
You always have a static array, behind a slice. However you can change the size of the slice, and go type system will kick in if you go off-bounds.

Slice is just a pointer to an array, with capacity and size, that's all. Two slices can share the same static array.

https://blog.golang.org/go-slices-usage-and-internals

Reply:
@nonsense I just added the lock; messing with atomics here required several calls, which in the end probably would end up contributing nothing more than obscurity.